### PR TITLE
stage2: implement most vector operations in Sema and LLVM backend

### PR DESCRIPTION
**Pull request by wsengir**
There was _a lot_ here, so I can't really go over it all line by line, but I did break up the commits into very neatly contained changes with some context in each commit message for the reader's benefit. At a high level:
* Add `cmp_vector` AIR instruction
* Make most `Value` functions auto-vectorize based on arguments
* Make a lot of Sema & LLVM backend vector-agnostic using auto-vectorizing
* Enable `test/behavior/vector.zig'

### CODE REVIEW

File: src/value.zig
Comment (andrewrk) [inline src/value.zig:2434]: Based on usage of this function it looks like the signature should instead be:

```suggestion
    pub fn compatibleVectorlikes(a: Value, b: Value) ?usize {
```

I see one usage where only 1 parameter is passed and it doesn't make sense to me to use this function at all in that case.

And then here this function signature reveals the same problems with `isVectorLike`; this is unsound without a `Type`.

File: src/value.zig
Comment (andrewrk) [inline src/value.zig:2468]: What does this do? This looks sus

**Review by andrewrk — CHANGES_REQUESTED** [review]
This is great work. It's also a lot of different stuff in one pull request, so naturally if *anything* in here is problematic, it's going to block the merge.

Up until 95d0753bd1f908187f09b75e136b42a3010e4876, the PR is flawless and can be merged immediately. However with that commit, I'm against merging it as is. I'm not necessarily entirely against moving vectorized operations from Sema to Value, but I am against it as implemented here, because it is unsound. I suggest that moving vectorized operation handling from Sema to Value happen in a separate change that we can evaluate on its own merits, for reasons that are hopefully clear with my review comments below.

As for 07737cc3d9e69e9c279338d6203214ed9d3c8817, I am against ever implementing this hack. Let's have a separate bug report for this issue and solve it in an appropriate way. I'm happy to look at it myself to unblock your work here. I'm also fine with changing the behavior tests to avoid triggering this bug, keeping the behavior tests limited to each only testing the one thing they are supposed to test.

File: src/Air.zig
Comment (wsengir) [reply src/Air.zig:949]: Whoops, sorry about that. I missed that when I paged through the diff. Will fix.

I really need to fix my editor setup so I can run `zig fmt` in real-time to catch these sooner. It really doesn't like auto formatting these huge files.

File: src/Air.zig
Comment (wsengir) [reply src/Air.zig:908]: Ah, I see now that these are treated almost identically. I'll leave this as-is for now using `bool_*`, but once I get this PR all fixed up, I can take a look at that!

File: src/value.zig
Comment (wsengir) [reply src/value.zig:1832]: This function doesn't create a new `Value` and only returns a `bool`, so we can't return a vector in its place. A lot of callsites use `compare` just as a simple test for "if this value is some known constant (i.e. an identity like 0 or 1), then I can do less work", but creating a vector & immediately reducing it with AND would be rather pointless and require needless allocations. The only time we would ever want to _not_ reduce vector comparisons with AND is when we're handling a Zig comparison instruction. Otherwise, we'll almost always just want a singular scalar to branch off of.

File: src/value.zig
Comment (wsengir) [reply src/value.zig:2421]: Thanks for explaining the policy on "unintuitive tricks that simplify things". I'll better document this (and make less hacky!).

File: src/value.zig
Comment (wsengir) [reply src/value.zig:2417]: I apologize for this. It was left from the more obtuse predecessor of `compatibleVectorlikes` and `indexVectorlike`. I'll delete this.

File: src/value.zig
Comment (wsengir) [reply src/value.zig:2468]: Which part, the return or `indexVectorlike`?

Regarding the return value, that's actually unnecessary now. Originally, `compatibleVectorlikes` had slightly different length logic & this is an artifact of that. As stated above, I'll be making the non-scalarized `Type` a required argument for each `Value` function to avoid needing to do any wacky "unsound" stuff here.

Regarding `indexVectorlike`, this is just a wrapper to avoid putting logic to handle `repeated` and implicitly vectorizable values (like `zero` and `one`) in every single vectorized function.

File: src/Air.zig
Comment (andrewrk) [reply src/Air.zig]: When you say it doesn't like it, what do you mean exactly? On my machine `zig fmt` takes 0.533s to format the entire standard library and self-hosted compiler. Perhaps you are using a debug build of the compiler for `zig fmt` instead of release?

File: src/value.zig
Comment (andrewrk) [reply src/value.zig:2414]: Apologies for the misclick- the comment was supposed to be only on the function signature.

If this function survives the other changes you are making, would you mind describing clearly what this function does in the doc comments?

**Comment by andrewrk** [comment]
> I didn't like it either, but I interpreted your comment regarding "work around it" in Sema as "keep the hack out of AstGen and we can come up with a permanent solution soon".

Ah I see. To be clear, I meant to work around it in the behavior test cases, avoiding triggering the bug. It's perfectly acceptable, and even desirable, for a bug to be covered by exactly one test case and have all the others avoid it.

Also I noticed that your comments here had the word "apologies" in them a couple times and I want to reiterate, your work is top caliber. You took a chance on a bold reworking, and got some guidance on how to complete it, and that's great. Nothing to apologize for here.

File: src/Air.zig
Comment (wsengir) [reply src/Air.zig]: I am indeed using a debug build, but `zig fmt` is by no means the bottleneck here. For whatever reason, Sublime Text takes a second or two to ponder the file update. VSCode handles it flawlessly, but weird Electron issues are preventing me from using it.

File: src/value.zig
Comment (wsengir) [reply src/value.zig:2414]: I just realized that I never responded to this comment. This is actually the only surviving function and I've added a descriptive doc comment to it.

**Pull request merged**
